src/note_encryption/domain.rs

```diff
@@ -271,6 +274,23 @@ impl<D: OrchardDomainCommon> Domain for OrchardDomain<D> {
         EphemeralSecretKey::from_bytes(out_plaintext.0[32..OUT_PLAINTEXT_SIZE].try_into().unwrap())
             .into()
     }
+
+    fn parse_note_plaintext_bytes(plaintext: &[u8]) -> Option<Self::NotePlaintextBytes> {
+        Self::NotePlaintextBytes::from_slice(plaintext)
+    }
+
+    fn parse_note_ciphertext_bytes(
+        output: &[u8],
+        tag: [u8; AEAD_TAG_SIZE],
+    ) -> Option<Self::NoteCiphertextBytes> {
+        Self::NoteCiphertextBytes::from_slice_with_tag(output, tag)
+    }
+
+    fn parse_compact_note_plaintext_bytes(
+        plaintext: &[u8],
+    ) -> Option<Self::CompactNotePlaintextBytes> {
+        Self::CompactNotePlaintextBytes::from_slice(plaintext)
+    }
```

Comment:
What is the point in adding this if not being used?

Reply:
The idea of introducing these functions came from [str4d's review comment](https://github.com/zcash/zcash_note_encryption/pull/2#discussion_r1652676085). Although these functions are not  used in the `orchard` or `sapling-crypto` crates directly, they are used in the `zcash_note_encryption` crate.

`zcash_note_encryption` delegates the implementation of these functions to upper-level crates, and then use them. 

Therefore, I added them to the `Domain` trait as methods and implemented them in both `orchard` and `sapling-crypto` crates.

However, as it turns out, the implementation is the same for both crates and relies solely on the `NoteBytes` trait methods. Given this, we can simplify the code by moving these function implementations directly to the `Domain` trait in the `zcash_note_encryption` crate.

I have tested this change, and it works correctly. I can proceed with moving the implementation (here and in `saplicg-crypto`) if everyone agrees.